src/datastore/access-policies.cpp

```diff
@@ -0,0 +1,152 @@
+#include "access-policies.h"
+
+#include <utility>
+
+#include <xid/xid.h>
+
+#include "err/errors.h"
+
+namespace datastore {
+AccessPolicy::AccessPolicy(const AccessPolicy::Data &data) noexcept : _data(data), _rev(0) {
+	if (_data.id.empty()) {
+		_data.id = xid::next();
+	}
+}
+
+AccessPolicy::AccessPolicy(Data &&data) noexcept : _data(std::move(data)), _rev(0) {
+	if (_data.id.empty()) {
+		_data.id = xid::next();
+	}
+}
+
+AccessPolicy::AccessPolicy(const pg::row_t &r) :
+	_data({
+		.id = r["_id"].as<std::string>(),
+	}),
+	_rev(r["_rev"].as<int>()) {}
+
+void AccessPolicy::store() const {
+	std::string_view qry = R"(
+		insert into "access-policies" as t (
+			_id,
+			_rev
+		) values (
+			$1::text,
+			$2::integer
+		)
+		on conflict (_id)
+		do nothing
+		returning _rev;
+	)";
+
+	pg::result_t res;
+	try {
+		res = pg::exec(qry, _data.id, _rev);
+	} catch (pqxx::check_violation &) {
+		throw err::DatastoreInvalidIdentityData();
+	} catch (pg::unique_violation_t &) {
+		throw err::DatastoreDuplicateIdentity();
+	}
+
+	if (res.empty()) {
+		throw err::DatastoreRevisionMismatch();
+	}
+
+	_rev = res.at(0, 0).as<int>();
+}
+
+void AccessPolicy::discard() const {
+	std::string_view qry = R"(
+		delete from "access-policies"
+		where
+			_id = $1::text;
+	)";
+
+	pg::exec(qry, _data.id);
+}
+
+void AccessPolicy::add_access(
+	const AccessPolicy::principal_t &principal, const AccessPolicy::resource_t &resource) const {
+	auto conn = datastore::redis::conn();
+	conn.cmd("SADD " + principal + ":" + resource + " " + _data.id);
```

Comment:
Thanks! The part where you talk about the key makes sense, but I'm a little confused about what you said about attributes.

My interpretation is that you want to add Redis hash records in addition to the sets that wil store attributes for each policy in the future (presumably so we don't have to query postgres). Is that correct?

Reply:
No, not in addition to the set, use a hash instead. e.g. `hset access:(cht592pjr928251v34ig)›[resource/xxxxxxxx] {{ policy id }} ""`. 

When we implement `CreateAccessPolicyRequest.rules[].attrs`, we'll store the JSON string of this in the hash against the policy ID instead of `""`.

Attributes are meant to add extra conditions for access, for example read only flag. So when we check access from an identity to a resource it's a single lookup in redis and in the return value we can send back all the attributes from different policies without having to do any other DB lookups.